source/common/network/filter_matcher.cc

```diff
@@ -0,0 +1,187 @@
+#include "common/network/filter_matcher.h"
+
+#include "envoy/network/filter.h"
+
+#include "absl/strings/str_format.h"
+
+namespace Envoy {
+namespace Network {
+
+// TwoPhaseMatcher is introduced to save number of memory allocation at build phase memory usage
+// match phase. Work with SetListenerFilterMatcher. The first phase is constructor, and the second
+// phase is finishBuild(). The invariance is that the direct matchers in a set logic matcher should
+// be continuous in the vector.
+//
+// Examples:
+//  `m := A && B && C` where `B := X || Y`
+// The 3 matchers "ABC" should be continuous, while the matchers "XY" should be continuous in the
+// storage of SetListenerFilterMatcher.
+class TwoPhaseMatcher : public ListenerFilterMatcher {
```

Comment:
In thinking about this more, can we just get rid of the two phase concept entirely? I don't think an inline implementation will use much more memory and it will be much easier to understand. The only reason the tap matcher works like this is because it needs to efficiently map a per-request matching state on top of a constant matching config. That doesn't apply here. Thank you!